tests/callbacks/test_inference.py

```diff
@@ -0,0 +1,101 @@
+from composer.models import composer_resnet
+from composer.trainer import Trainer
+from tests.common.datasets import RandomImageDataset
+
+
+@pytest.mark.parametrize(
+    'model_cls, sample_input',
+    [
+        (partial(composer_resnet, 'resnet18'), (torch.rand(1, 3, 224, 224), torch.randint(10, (1,)))),
+    ],
+)
+def test_inference_callback_torchscript(model_cls, sample_input):
+    save_format = 'torchscript'
+    model = model_cls()
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        save_path = os.path.join(tempdir, f'model.pt')
+        export_for_inference = ExportForInference(save_format=save_format, out_filename=str(save_path))
+
+        # Construct the trainer and train
+        trainer = Trainer(
+            model=model,
+            callbacks=export_for_inference,
+            train_dataloader=DataLoader(RandomImageDataset(shape=(3, 224, 224))),
+            max_duration='1ba',
+        )
+        trainer.fit()
+
+        model.eval()
+        orig_out = model(sample_input)
+
+        loaded_model = torch.jit.load(save_path)
+        loaded_model.eval()
+        loaded_model_out = loaded_model(sample_input)
+
+        assert torch.allclose(
+            orig_out,
+            loaded_model_out), f'mismatch in the original and exported for inference model outputs with {save_format}'
+
+
+@pytest.mark.parametrize(
+    'model_cls, sample_input',
+    [
+        (partial(composer_resnet, 'resnet18'), (torch.rand(1, 3, 224, 224), torch.randint(10, (1,)))),
+    ],
+)
+def test_inference_callback_onnx(model_cls, sample_input):
+    pytest.importorskip('onnx')
+    pytest.importorskip('onnxruntime')
+    import onnx  # type: ignore
+    import onnxruntime as ort  # type: ignore
+
+    save_format = 'onnx'
+    model = model_cls()
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        save_path = os.path.join(tempdir, f'model.onnx')
+        export_for_inference = ExportForInference(save_format=save_format, out_filename=str(save_path))
+
+        # Construct the trainer and train
+        trainer = Trainer(
+            model=model,
+            callbacks=export_for_inference,
+            train_dataloader=DataLoader(RandomImageDataset(shape=(3, 224, 224))),
+            max_duration='1ba',
+        )
+        trainer.fit()
+
+        model.eval()
+        orig_out = model(sample_input)
+
+        loaded_model = onnx.load(save_path)
+        onnx.checker.check_model(loaded_model)
+
+        ort_session = ort.InferenceSession(save_path)
+        loaded_model_out = ort_session.run(
+            None,
+            {'input': sample_input[0].numpy()},
+        )
+
+        torch.testing.assert_close(
+            orig_out.detach().numpy(),
+            loaded_model_out[0],
+            rtol=1e-4,  # lower tolerance for ONNX
+            atol=1e-3,  # lower tolerance for ONNX
+            msg=f'mismatch in the original and exported for inference model outputs with {save_format}')
```

Comment:
msg is probably not needed, as pytest errors will already have descriptions attached.